core/commonMain/src/kotlinx/serialization/internal/ElementMarker.kt

```diff
@@ -0,0 +1,93 @@
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.InternalSerializationApi
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.CompositeDecoder
+
+@InternalSerializationApi
+@OptIn(ExperimentalSerializationApi::class)
+public abstract class ElementMarker(private val descriptor: SerialDescriptor) {
+    /*
+    Element decoding marks from given bytes.
+    The element number is the same as the bit position.
+    Marks for the lowest 64 elements are always stored in a single Long value, higher elements stores in long array.
+     */
+    private var lowerMarks: Long
+    private val highMarksArray: LongArray?
+
+    init {
+        val elementsCount = descriptor.elementsCount
+        if (elementsCount <= Long.SIZE_BITS) {
+            lowerMarks = if (elementsCount == Long.SIZE_BITS) {
+                // number of bits in the mark is equal to the number of fields
+                0L
+            } else {
+                // (1 - elementsCount) bits are always 1 since there are no fields for them
+                -1L shl elementsCount
+            }
+            highMarksArray = null
```

Comment:
[nit] you can declare an empty array somewhere and assign it here.
Then you can avoid `!!` in the code